File: articles/linear-algebra-for-systems-modeling/articles/span-linear-independence-and-basis/cpp/span_basis_audit.cpp
#include <array>
#include <cmath>
#include <iostream>

double determinant3x3(const std::array<std::array<double, 3>, 3>& m) {
    return m[0][0]*(m[1][1]*m[2][2] - m[1][2]*m[2][1])
         - m[0][1]*(m[1][0]*m[2][2] - m[1][2]*m[2][0])
         + m[0][2]*(m[1][0]*m[2][1] - m[1][1]*m[2][0]);
}

int main() {
    std::array<std::array<double, 3>, 3> matrix {{
        {{1.0, 0.0, 0.5}},
        {{0.0, 1.0, 0.5}},
        {{0.0, 0.0, 1.0}}
    }};

    int ambient_dimension = 3;
    int vector_count = 3;
    double det = determinant3x3(matrix);
    int rank = std::abs(det) > 1e-10 ? 3 : 2;
    bool spans = rank == ambient_dimension;
    bool independent = rank == vector_count;
    bool basis = spans && independent;

    std::cout << "vector_set_name,ambient_dimension,vector_count,rank,spans_ambient_space,linearly_independent,is_basis_for_ambient_space,warning\n";
    std::cout << "candidate_system_basis,3,3," << rank << "," << spans << "," << independent << "," << basis
              << ",A mathematical basis claim does not prove real-world adequacy.\n";
    return 0;
}

File: articles/linear-algebra-for-systems-modeling/articles/span-linear-independence-and-basis/julia/span_basis_audit.jl
function determinant3x3(m)
    return m[1,1]*(m[2,2]*m[3,3] - m[2,3]*m[3,2]) -
           m[1,2]*(m[2,1]*m[3,3] - m[2,3]*m[3,1]) +
           m[1,3]*(m[2,1]*m[3,2] - m[2,2]*m[3,1])
end

matrix = [
    1.0 0.0 0.5;
    0.0 1.0 0.5;
    0.0 0.0 1.0
]

ambient_dimension = 3
vector_count = 3
det_value = determinant3x3(matrix)
rank_value = abs(det_value) > 1e-10 ? 3 : 2
spans = rank_value == ambient_dimension
independent = rank_value == vector_count
basis = spans && independent

println("vector_set_name,ambient_dimension,vector_count,rank,spans_ambient_space,linearly_independent,is_basis_for_ambient_space,warning")
println(join((
    "candidate_system_basis",
    ambient_dimension,
    vector_count,
    rank_value,
    spans,
    independent,
    basis,
    "A mathematical basis claim does not prove real-world adequacy."
), ","))
